klient/k8s/yaml/yaml.go

```diff
@@ -0,0 +1,130 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package yaml
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"path/filepath"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/util/yaml"
+	"sigs.k8s.io/e2e-framework/klient"
+	"sigs.k8s.io/e2e-framework/klient/k8s/resources"
+)
+
+// LoadUnstructuredYAML loads a single-document YAML or JSON file
+func LoadUnstructuredYAML(filename string) (*unstructured.Unstructured, error) {
```

Comment:
Idea:
```suggestion
func LoadUnstructured(filename string) (*unstructured.Unstructured, error) {
```

Reply:
Another thought on the signature of this:
```suggestion
func LoadResource(filename string, obj k8s.Object) (error) {
```
Allowing users to pass in a typed object might provide for a better experience.

Reply:
And for a final take at generic support -- this would support files (os.Open), strings (strings.NewReader), or bytes (bytes.NewReader) with any typed object:
```suggestion
func LoadResource(file io.Reader, obj k8s.Object) (error) {
```

Reply:
Personally, I like the last approach better. This can also avoid you having to create temp files just to call the load method in case if you are generating the template using a `text/template` or some other high level template infra. And the loading from the dir can proxy off to this same method like you have done already.

Reply:
And the function names can be `Load` and `LoadFromDir` possibly ?

```go
package mypkg

import (
    "testing"
    "sigs.k8s.io/e2e-framework/klient/k8s/resources"
)

func TestSomething(t *testing.T) {
    // do things here to create a io.Reader instance
    _ = resources.Load(reader, &pod)
    // do things with it.
    _ = resources.LoadFromDir(podManifestsDir, &pods)
}
```

Reply:
One downside to the `(manifest io.Reader, obj k8s.ObjectList)` signature is that if your directory has different types, you have to know of a permissive-list type (e.g. UnstructuredList) that would be applicable

Reply:
> One downside to the (directory string, obj k8s.ObjectList) signature is that if your directory has different types, you have to know of a permissive-list type 

That is a very valid thing you brought up. The advantage of being able to reuse `k8s.Object` is you can easily get the typing support without having to worry about casting it. But using `Unstructured` makes the reading and un-marshalling much easier. 

If you weigh functionality vs the ease of typing, I would personally pick the functionality in this case. Let us see what @vladimirvivien  or @ShwethaKumbla have to say about this behavior.

Reply:
> If you weigh functionality (unstructured) vs the ease of typing (k8s.Object), I would personally pick the functionality in this case.

I think we can get both, following some of the patterns from the conditional package; k8s.Object for generic / typed use-cases and helpers that use Unstructured by default for the 99% / simple use-cases (that wrap the k8s.Object-based functions).